src/runtime_src/core/edge/sw_em/generic_pcie_hal2/shim.cxx

```diff
@@ -214,10 +214,11 @@ namespace xclcpuemhal2 {
     }
     if (buf_size < new_size)
     {
-      void *temp = buf;
-      buf = (void*) realloc(buf,new_size);
-      if (!buf) // prevent leak of original buf
-        free(temp);
+      void *result = realloc(buf, new_size);
+      // If realloc was unsuccessful, then return the original size of buf.
+      if (result == NULL)
+        return buf_size;
+      buf = result;
```

Comment:
It is unclear how bad allocs are handled in this framework at the first place.
Perhaps by just setting `buf` to `nullptr` and thus freeing the memory sounds good too.
In your case you detect the bad alloc but you go on with a too small buffer, which might cause havoc too...

Reply:
Disclaimer: I am new to XRT.

> ... thus freeing the memory sounds good too.
What is the purpose of `CpuemShim::alloc_void()`? In my view it uses an already allocated buffer or enlarges it if required. It cannot get any smaller but only larger. In some cases there is not enough memory, then I see two approaches:
1. Wait and try again
2. Error out

>  you go on with a too small buffer, which might cause havoc too...
My solution gives the caller the opportunity to execute the 1. approach.

Freeing the memory means giving up, so we could error out immediately. `realloc()` would free up the memory anyway if it can enlarge the existing block or finds a new larger block anyway. What do you think?

Reply:
All this emulation shim in XRT is to be rewritten at some point because it does not meet some basic quality criteria and has a lot of technical debt, so I suggest to not spending too much time on this (if you are really motivated AMD is hiring a lot ;-) ).
We cannot do much when there is not enough memory anymore.
The caller might not do what you expect.
So, why not just make the code to compile?

Reply:
To be clearer, what about also discarding the old buffer if the reallocation failed to get bigger memory as it was done before?
At the end the allocation is not usable anymore anyway since there is not enough memory, so why not giving it back to the community?

Reply:
> ... so I suggest to not spending too much time on this
Thanks, good to know this 🙂

> why not giving it back to the community?
done

> So, why not just make the code to compile?
I thought the `realloc` code could lead to problems. Currently the code compiles but if you think these fixes are needless due to a planned rewrite, feel free to close this PR 😉

Actually my goal is to get XRT running with an Alveo on my development machine with a recent kernel. Today I found out that the kernel driver included in this repo is for older kernels. The [new version seems to be on the kernel pipeline](https://patchwork.kernel.org/project/linux-fpga/patch/20220105225013.1567871-2-lizhi.hou@xilinx.com/), so maybe this repo won't work with the new driver at all.

Reply:
All thwse might take quite some time. So, just use this version in the meantime.
I am unsure the CI tests the case when there is no more memory.
You might try with `ulimit`. :-)